protocol-units/da-sequencer/node/src/celestia/mod.rs

```diff
@@ -1,93 +1,160 @@
+		CelestiaExternalDa { notifier, celestia_client }
 	}
 
-	/// send the given block to Celestia.
-	/// The block is not immediatly sent but aggergated in a blob
-	/// Until the client can send it to celestia.
-	pub async fn send_block(&self, block: &SequencerBlockDigest) -> Result<(), DaSequencerError> {
+	async fn request_block(&self, at: BlockAt) -> Result<SequencerBlock, DaSequencerError> {
+		let (tx, rx) = oneshot::channel();
+		let request = ExternalDaNotification::RequestBlock { at, callback: tx };
+		self.notifier.send(request).await.map_err(|e| {
+			DaSequencerError::BlockRetrieval(format!("Broken notifier channel: {}", e))
+		})?;
+		let block = rx.await.map_err(|e| {
+			DaSequencerError::BlockRetrieval(format!("Broken notifier channel: {}", e))
+		})?;
+		let block = block
+			.ok_or(DaSequencerError::BlockRetrieval(format!("Block at {:?} not found", at)))?;
+		Ok(block)
+	}
+}
+
+impl<C: CelestiaClient + Sync + Clone> DaSequencerExternalDa for CelestiaExternalDa<C> {
+	/// Send the given block to Celestia.
+	/// The block is not immediately sent but aggregated in a blob
+	/// until the client can send it to celestia.
+	async fn send_block(&self, _block: SequencerBlockDigest) -> Result<(), DaSequencerError> {
 		todo!()
 	}
 
 	/// Get the blob from celestia at the given height.
-	pub async fn get_blob_at_height(&self) -> Result<Option<Blob>, DaSequencerError> {
+	async fn get_blobs_at_height(
+		&self,
+		_height: CelestiaHeight,
+	) -> Result<Option<Vec<Blob>>, DaSequencerError> {
 		todo!()
 	}
 
-	/// Bootstrap the Celestia client to recovert from missing block.
+	/// Bootstrap the Celestia client to recover from missing block.
 	/// In case of crash for example, block sent to Celestia can be behind the block created by the network.
-	/// The role of this function is to recovert this missing block to all block of the network are sent to celestia.
-	/// The basic algorythm is start from 'last_sent_block_height' to 'current_block_height' and request using the notifier channel
-	/// The missing block. Not sure last_notified_celestia_height is useful.
+	/// The role of this function is to recover this missing block to all block of the network are sent to celestia.
+	/// The basic algorithm is start from 'last_sent_block_height' to 'current_block_height' and request using the notifier channel
+	/// the missing block. Not sure last_notified_celestia_height is useful.
 	/// During this boostrap new block are sent to the client.
 	/// These block should be buffered until the boostrap is done then sent after in order.
-	pub async fn bootstrap(
+	async fn bootstrap(
 		&self,
 		current_block_height: BlockHeight,
-		last_sent_block_height: BlockHeight,
-		last_notified_celestia_height: CelestiaHeight,
+		last_finalized_celestia_height: CelestiaHeight,
 	) -> Result<(), DaSequencerError> {
-		todo!()
+		// wait to ensure that no blob is pending in the Celestia network
+		tokio::time::sleep(DELAY_SECONDS_BEFORE_BOOTSTRAPPING).await;
```

Comment:
During this wait, new blobs can arrive, so you need to update the `last_finalized_celestia_height` if any.
To get this info you can add a call to the Celestia client, a sort of notification for new blob for example. See with @mzabaluev the best way to be updated.

Reply:
This makes sense. I will remove the `last_finalized_celestia_height` parameter from this function and require the `CelestiaClient` to provide a method for retrieving the current celestia height.

Reply:
You need this parameter in case there's no pending blob, so keep it. I just say that you need to get new committed blob during that wait. If there's any they will update this parameter.